server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java

```diff
@@ -404,7 +404,7 @@ public void addRescore(RescoreContext rescore) {
 
     @Override
     public boolean hasScriptFields() {
-        return scriptFields != null;
+        return scriptFields != null && scriptFields.fields().isEmpty() == false;
```

Comment:
This was a fun one to track down.  If it's not here, then once you have called `scriptFields()` once, `hasScriptFields` always returns `true`, because `scriptFields()` creates an empty instance.  Remember folks, idempotency is a Good Thing and prevents many hours of head scratching.